### PR TITLE
Gnss sim

### PR DESCRIPTION
**Pull request by superjax**
Simualted GNSS measurements in gazebo

### CODE REVIEW

**Comment by superjax** [comment]
This needs to be tested in Sim.

File: rosflight_sim/src/sil_board.cpp [hunk 1]
```diff
@@ -46,9 +46,17 @@ void SIL_Board::init_board(void)
   boot_time_ = GZ_COMPAT_GET_SIM_TIME(world_);
 }
 
-void SIL_Board::gazebo_setup(gazebo::physics::LinkPtr link, gazebo::physics::WorldPtr world,
-                             gazebo::physics::ModelPtr model, ros::NodeHandle* nh, std::string mav_type)
+constexpr double rad2Deg(double x)
+{
+  return M_PI/180.0 * x;
```
Comment (plusk01) [inline rosflight_sim/src/sil_board.cpp]: shouldn't this be `180.0/M_PI * x`?

File: rosflight_sim/src/sil_board.cpp [hunk 2]
```diff
@@ -46,9 +46,17 @@ void SIL_Board::init_board(void)
...
+}
+constexpr double deg2Rad(double x)
 {
+  return 180.0/M_PI * x;
```
Comment (plusk01) [inline rosflight_sim/src/sil_board.cpp]: shouldn't this be `M_PI/180.0 * x`?

File: rosflight_sim/src/sil_board.cpp [hunk 3]
```diff
@@ -490,23 +505,155 @@ void SIL_Board::RCCallback(const rosflight_msgs::RCRaw& msg)
...
+  // We are going to pretend the earth is perfectly round
+  static constexpr double R = 6378145.0f;
+
+  rosflight_firmware::GNSSData out;
+  out.fix_type = rosflight_firmware::GNSSFixType::GNSS_FIX_TYPE_FIX;
+  out.time_of_week = GZ_COMPAT_GET_SIM_TIME(world_).Double()*1000;
+  out.time = GZ_COMPAT_GET_SIM_TIME(world_).Double();
+  out.nanos = (GZ_COMPAT_GET_SIM_TIME(world_).Double() - out.time)*1e9;
+
+  GazeboPose current_state_NWU = GZ_COMPAT_GET_WORLD_POSE(link_);
```
Comment (plusk01) [inline rosflight_sim/src/sil_board.cpp]: What is NWU? That coordinate frame doesn't exist... :-)

https://www.ros.org/reps/rep-0103.html#id19

**Comment by BillThePlatypus** [comment]
@plusk01 LLA is provided for convenience. In particular, the standard ROS message type for GNSS information is [sensor_msgs/NavsatFix](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/NavSatFix.html). Although here at the lab we mostly use ECEF, this offers others compatibility with third party packages.

**Comment by BillThePlatypus** [comment]
It would probably be best to use Gazebo's native coordinate transformation system. I used it for the rosplane_sim GPS plugin when we added GNSS.

One feature I think would be good is using the spherical coordinate provided by the Gazebo World for the origin. In the Gazebo world file, you can specify a LLA for the origin. By using this throughout, it means that multiple plugins can draw from the same source, for example the GPS plugin and the ROSplane truth plugin.

**Comment by plusk01** [comment]
Thanks @BillThePlatypus; I should have been more specific: the way the code is written (calc ECEF from LLA) makes LLA seem like a first-class citizen, but the argument is that ECEF should be the underlying representation and then LLA should be calculated at the end for compatibility.

Just curious about the implications of this approach (specifically accuracy).

**Comment by BillThePlatypus** [comment]
One thing to note is that while GPS may use ECEF in the background, the noise in the output is different in the horizontal and the vertical directions. This means that there will have to be some coordinate transform, at the very least from local NED to ECEF.

**Comment by superjax** [comment]
Thanks for pointing out the `GpsSensor`!  That is probably a lot better.  I'll look into it.

As for why lla seems to be a first-class citizen, it's because I was tired.  I'm glad you pointed it out, because I think that migrating to the GpsSensor makes a lot more sense.  I started out using some utilities for handling WGS84 stuff, [link](https://github.com/superjax/gnss_utils/blob/master/include/gnss_utils/wgs84.h), but bringing that in ended up in a [banana-gorilla-jungle problem](https://www.defprogramming.com/quotes-by/joe-armstrong/).

However, from an API perspective, I have found that specifying lla origin tends to be a lot more intuitive than ECEF.  Kinda like specifying RPY for sensor orientations is a lot easier than a quaternion or rotation matrix.  I think I'll keep the `origin_` parameters in `lla` unless there are strong objections.

I'll re-do this with one of those APIs and that will clean this all up.  I suspect that the `GpsSensor` also models noise a lot better than I am in this PR.

File: rosflight_sim/src/sil_board.cpp (hunk 3 above)
Comment (superjax) [reply rosflight_sim/src/sil_board.cpp]: Okay, okay I'll switch to ENU.  But not this commit.  :)  

Let's wait for this to land, and then let's have another (or maybe just continue our discussion about it)

**Comment by superjax** [comment]
Just to close the loop on the discussion.  I dug into the `GpsSensor`, and decided against it because it forces us to use the `sdf` (or `xacro` in our case) to define parameters like the origin and noise characteristics.  Also, they weren't doing anything terribly fancy with the noise characteristics, so there wasn't a whole lot of advantage to using it anyway.

Everything else in `rosflight_sim` uses the ROS parameter server to configure the plugins because that's a lot simpler than having two separate parameter configuration systems.  Anyway, Spherical Coordinates were a lot less integrated into the gazebo ecosystem.

A future todo would be to improve the GPS noise modeling.

**Comment by BillThePlatypus** [comment]
I didn't intend for this to get merged now. It depends on battery monitor
code that hasn't yet been merged. Also, did you get the slight airbourne
changes that this is dependent on? The documentation also needs some
changes.

**Comment by BillThePlatypus** [comment]
Ignore my last comment. I got PR's confused.

**Pull request merged**